Sources/SwiftProtobuf/BinaryEncodingSizeVisitor.swift

```diff
@@ -18,54 +18,195 @@ import Foundation
 /// Visitor that calculates the binary-encoded size of a message so that a
 /// properly sized `Data` or `UInt8` array can be pre-allocated before
 /// serialization.
-final class BinaryEncodingSizeVisitor: Visitor {
+internal struct BinaryEncodingSizeVisitor: Visitor {
 
   /// Accumulates the required size of the message during traversal.
   var serializedSize: Int = 0
 
   init() {}
 
-  func visitUnknown(bytes: Data) {
+  mutating func visitUnknown(bytes: Data) {
     serializedSize += bytes.count
   }
 
-  func visitSingularField<S: FieldType>(fieldType: S.Type,
-                                        value: S.BaseType,
-                                        fieldNumber: Int) throws {
-    let tagSize = FieldTag(fieldNumber: fieldNumber,
-                           wireFormat: S.protobufWireFormat).encodedSize
-    serializedSize += try tagSize + S.encodedSizeWithoutTag(of: value)
+  public mutating func visitSingularFloatField(value: Float, fieldNumber: Int) throws {
+    let tagSize = FieldTag(fieldNumber: fieldNumber, wireFormat: .fixed32).encodedSize
+    serializedSize += tagSize + MemoryLayout<Float>.size
   }
 
-  func visitRepeatedField<S: FieldType>(fieldType: S.Type,
-                                        value: [S.BaseType],
-                                        fieldNumber: Int) throws {
-    let tagSize = FieldTag(fieldNumber: fieldNumber,
-                           wireFormat: S.protobufWireFormat).encodedSize
-    serializedSize += value.count * tagSize
+  public mutating func visitSingularDoubleField(value: Double, fieldNumber: Int) throws {
```

Comment:
This is all quite repetitive but at least it's pretty easy to read and understand -- the verbosity here is in exchange for dropping the various helper methods off of the `FieldType`s.